### ISSUE TITLE
Tooltip is displayed incorrectly when the parent container is transformed with CSS "transform: scale()" property

### ISSUE THREAD

**Issue by MateuszTrN** [recharts/recharts#2778]
## Do you want to request a feature or report a bug?
Bug

## What is the current behavior?
If the parent (or any container above) div is scaled via css `transform: scale(x)` property - the mouse events work incorrectly. 
The tooltip is displayed for the data that is not hovered. Trying to display tooltip for the few last bars is impossible. 

## If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via 
https://codesandbox.io/s/simple-bar-chart-forked-enk3g?file=/src/App.tsx

## What is the expected behavior?
regardless of parent div transformation, the mouse events / tooltip work the same

## Which versions of Recharts, and which browser / OS are affected by this issue? Did this work in previous versions of Recharts?
Recharts 2.1.8 (I quickly tested previous versions  1.x, 2.0.x - same behaviour)
OS: MacOS Mojave
Browser Chrome Version 97.0.4692.99 (Official Build) (x86_64)

**Comment by MateuszTrN** [comment]
Hi,
I did some digging and I found the cause of the issue. I applied some quickfixes to ensure if my thinking was right and it seems to be working with those changes. See below:

file: ` ../chart/getCategoricalChart.js ` lines `1209-1229`
Original code snippet
```
inRange(x: number, y: number): any {
      const { layout } = this.props;

      if (layout === 'horizontal' || layout === 'vertical') {
        const { offset } = this.state;
        const isInRange =
          x >= offset.left && x <= offset.left + offset.width && y >= offset.top && y <= offset.top + offset.height;

        return isInRange ? { x, y } : null;
      }

      const { angleAxisMap, radiusAxisMap } = this.state;

      if (angleAxisMap && radiusAxisMap) {
        const angleAxis = getAnyElementOfObject(angleAxisMap);

        return inRangeOfSector({ x, y }, angleAxis);
      }

      return null;
    }
```


my hotfix:
```
  inRange(x: number, y: number): any {
      const { layout } = this.props; 
      //calculating scale
      const element = this.container; 
     
     
       //the line below is the most dirty part of this and I wouldn't implement it here in final solution. 
       const scale = element.getBoundingClientRect().width / element.offsetWidth;
       //calculating scaled coordinates
        const [scaledX, scaledY] = [x / scale, y / scale]; 
       
      if (layout === 'horizontal' || layout === 'vertical') {    
        const { offset } = this.state;
     
        //replace x,y with the scaled counterparts 
        const isInRange =
          scaledX >= offset.left && scaledX <= offset.left + offset.width && scaledY >= offset.top && scaledY <= offset.top + offset.height;

        return isInRange ? { scaledX, scaledY } : null;
      }

      const { angleAxisMap, radiusAxisMap } = this.state;

      if (angleAxisMap && radiusAxisMap) {
        const angleAxis = getAnyElementOfObject(angleAxisMap);
        //it would be needed to consider this part as well, but I haven't checked that out yet.
        return inRangeOfSector({ x, y }, angleAxis);
      }

      return null;
    }

```

If I'd implement the fix I'd move the scale calculation into the event function and pass it as an default arg to the inRange function something like: 

```
 getMouseInfo(event: any) {
      if (!this.container) {
        return null;
      }
      
       //calculating scale
      const element = this.container; 
      const scale = element.getBoundingClientRect().width / element.offsetWidth;

      const containerOffset = getOffset(this.container);
      const e = calculateChartCoordinate(event, containerOffset);
      const rangeObj = this.inRange(e.chartX, e.chartY, scale);
      if (!rangeObj) {
        return null;
      }
```
and than:
```
  inRange(x: number, y: number, scale = 1): any {
      const { layout } = this.props; 
//... and the rest more or less the same
```

**Comment by norayr93** [comment]
@MateuszTrN  Can you please describe where are those functions put ? Seems like these are just custom function names.

**Comment by MateuszTrN** [comment]
> @MateuszTrN Can you please describe where are those functions put ? Seems like these are just custom function names.

Hi, 
Because this issue wasn't addressed by the authors of recharts, I ended up with a hack using webpack NormalModuleReplacementPlugin and I am hacking recharts compiled code on the fly when the webpack is building the whole app.  
I'm pretty sure it's not the best solution, but I didn't have time and budget to come up with anything better.  This is the function that we replaced 

https://github.com/recharts/recharts/blob/master/src/chart/generateCategoricalChart.tsx#L1209

Be aware that this hack doesn't fix all types of charts - because I didn't mess with `inRangeOfSector` function

(note that this is a tsx code and the webpack NormalModuleReplacementPlugin is pointed to complied *.js file from `/node_modules/recharts/es6/chart/generateCategoricalChart.js`):
the js code that we use: 
`{
      key: "inRange",
      value: function inRange(x, y) {
        var layout = this.props.layout;

        if (layout === 'horizontal' || layout === 'vertical') {
          var element = this.container;
          var scale = element.getBoundingClientRect().width / element.offsetWidth;
          var scaled = [x / scale, y / scale];
          var scaledX = scaled[0];
          var scaledY = scaled[1];
          var offset = this.state.offset;
          var isInRange =
            scaledX >= offset.left && scaledX <= offset.left + offset.width && scaledY >= offset.top && scaledY <= offset.top + offset.height;

          return isInRange ? { x: scaledX, y: scaledY } : null;
        }

        var _this$state12 = this.state,
            angleAxisMap = _this$state12.angleAxisMap,
            radiusAxisMap = _this$state12.radiusAxisMap;

        if (angleAxisMap && radiusAxisMap) {
          var angleAxis = getAnyElementOfObject(angleAxisMap);
          return inRangeOfSector({
            x: x,
            y: y
          }, angleAxis);
        }

        return null;
      }
    }`

**Comment by zhonglin94** [comment]
I met the same problem

**Comment by nikolasrieble** [comment]
@MateuszTrN Would you like to fix this in recharts? 
I can also add your solution, but you have already done all the work. 

If it helps, I can add a story to storybook, replicating the issue, which we can then use to test the solution as well.

**Comment by MateuszTrN** [comment]
> @MateuszTrN Would you like to fix this in recharts? I can also add your solution, but you have already done all the work.
> 
> If it helps, I can add a story to storybook, replicating the issue, which we can then use to test the solution as well.

Hi, I'm happy to help if I can. I'll get back to you in few days as I'm a bit busy in this week

**Comment by ckifer** [comment]
Released in 2.9, closing!
